### ISSUE TITLE
Setup multi-tenant auth

### ISSUE THREAD

**Issue by pcr0wley** [crowleysystems/dittofeed#2]
Dittofeed currently uses single-tenant auth, which really just means that there is a single user account.

In their code I found support for multiple workspaces and multiple user accounts.

Additionally they have a `multi-user` mode that allows auth via OIDC.

https://github.com/dittofeed/dittofeed/blob/fa439ccd69e96a4b01b6254fb17b8365fd4ae559/packages/backend-lib/src/config.ts#L58

```
OIDC_TOKEN_PUBLIC_KEY: oidcTokenPublicKey
AUTH_MODE: authMode
AUTH_PROVIDER: authProvider
OAUTH_START_URL: oauthStartUrl
SIGNOUT_URL: signoutUrl
```

Max at Dittofeed did mention that they need to use some sort of proxy on Kubernetes to handle auth. He also said that he would make that info available to us so that we can setup multi-tenant auth. _We do still need to see how they handle multiple workspaces and roles._

On a side note I think Dittofeed basically just setup a large infrastructure to handle many tenants at scale, and then opensourced the code. It isn't really user friendly for smaller companies, but I think in the end the infrastructure will help as we build out our product.

**Comment by pcr0wley** [comment]
I spoke with Max and it seems that multi-tenant documentation definitely isn't a priority at the moment. I imagine not only are they busy, but also hesitant to document the last remaining feature that distinguishes their cloud offering from the opensource one.

In any case I looked through the code and did a bit more research and it seems that Dittofeed relies on the OAuth2 Proxy for everything.

With multi-tenant setup then all traffic would flow through the proxy, which would make sure it was authenticated. I am not sure if the proxy would then pass on a header to Dittofeed, or if once authenticated all traffic would flow as normal.

Since there would not be any authorization between the proxy and Dittofeed we could probably just use Odoo for authentication. It would not necessarily need to be OAuth2.

**Comment by pcr0wley** [comment]
This is not a priority at the moment. First we need to figure out our overall authentication scheme. Namely whether we would use Odoo for authentication across all apps, or a third-party auth provider like Google, Auth0, etc. Or multiple providers.

### Auth with multiple providers
<img width="697" alt="Screenshot 2023-12-13 at 2 46 48 PM" src="https://github.com/crowleysystems/dittofeed/assets/913268/cf171082-2ba3-4c30-b1b8-26e49601b895">

**Comment by pcr0wley** [comment]
> I spoke with Max and it seems that multi-tenant documentation definitely isn't a priority at the moment. I imagine not only are they busy, but also hesitant to document the last remaining feature that distinguishes their cloud offering from the opensource one.

I spoke with Max again and this is definitely the case. I don't see them releasing multi-tenant documentation any time soon.

Once we have time however we can go through the code and figure out how to manage Auth. Since it uses a proxy we would just need to create an app in Odoo to replicate the proxy. As the code is already setup to allow a third-party to create new workspaces and users things should automatically work once we get auth figured out. That is to say we should be able to easily create workspaces and users per Odoo instance.

**Comment by pcr0wley** [comment]
Max at Dittofeed will be providing Kubernetes info and Helm charts. Though we won't use it as-is it should be helpful in getting this setup.

As far as I can tell when `multi-tenant` mode is setup Dittofeed works with multiple workspaces and users, but is not secured. The workspace and user would be communicated somehow in either a JWT or session cookie.

`[oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy)` would accept all requests, authorize them and then send them through to Dittofeed.

Basically this means that we have three options for multi-tenant functionality:
1. Setup `oauth2-proxy` and an identity provider
2. Create a proxy in Odoo and handle authorization there
3. Modify the Dittofeed code to integrate with Odoo auth

**Comment by pcr0wley** [comment]
Below is info from Max at Dittofeed.

## Ingress controller config

Install the ingress-nginx ingress controller.

```bash
helm install ingress-nginx ingress-nginx/ingress-nginx
```

## oauth-proxy config

Install oauth2-proxy for oauth with nginx reverse proxy.

```bash
helm upgrade --install -f k8s-infra/oauth2-proxy-cloud.yaml dittofeed-cloud-oauth2 oauth2-proxy/oauth2-proxy
```

```yaml
config:
  existingSecret: oauth2-proxy-creds

extraArgs:
  whitelist-domain: "dittofeed.com,.us.auth0.com"
  cookie-domain: "dittofeed.com,.us.auth0.com"
  provider: "oidc"
  ping-user-agent: "GoogleHC/1.0"
  silence-ping-logging: "true"
  email-domain: "*"
  cookie-refresh: "1h"
  oidc-issuer-url: "https://dittofeed.us.auth0.com/"
  client-id: "my-client-id"
  client-secret: "$(OAUTH2_PROXY_CLIENT_SECRET)"
  set-authorization-header: "true"
  redirect-url: "https://dittofeed.com/oauth2/callback"
  cookie-secret: "$(OAUTH2_PROXY_COOKIE_SECRET)"
  code-challenge-method: "S256"
  insecure-oidc-allow-unverified-email: "true"

ingress:
  enabled: true
  path: /oauth2
  pathType: Prefix
  hosts:
    - dittofeed.com
  className: nginx
  annotations:
    acme.cert-manager.io/http01-edit-in-place: "true"
    cert-manager.io/cluster-issuer: "letsencrypt-prod"
  tls:
    - secretName: oauth2-proxy-cloud-https-cert
      hosts:
        - dittofeed.com

extraEnv:
  - name: OAUTH2_PROXY_CLIENT_SECRET
    valueFrom:
      secretKeyRef:
        name: auth0-prod-web
        key: client-secret
  - name: OAUTH2_PROXY_COOKIE_SECRET
    valueFrom:
      secretKeyRef:
        name: oauth2-proxy-cookie-secret
        key: cookie-secret
```

## Ingress Resources

Create two Ingress resources, one for public endpoints, the other for oauth authenticated endpoints.

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: marketing-automation
  namespace: marketing-automation
spec:
  ingressClassName: nginx
  rules:
    - host: dittofeed.com
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: marketing-automation-site
                port:
                  number: 80
          - path: "/api/public"
            pathType: Prefix
            backend:
              service:
                name: dittofeed-cloud-api
                port:
                  number: 80
          - path: "/dashboard/public"
            pathType: Prefix
            backend:
              service:
                name: dittofeed-cloud-dashboard
                port:
                  number: 80
          - path: "/dashboard/404"
            pathType: Prefix
            backend:
              service:
                name: dittofeed-cloud-dashboard
                port:
                  number: 80
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: marketing-automation-auth
  namespace: marketing-automation
  annotations:
    nginx.ingress.kubernetes.io/auth-url: "https://$host/oauth2/auth"
    nginx.ingress.kubernetes.io/auth-signin: "https://$host/oauth2/start?rd=$escaped_request_uri"
    nginx.ingress.kubernetes.io/auth-response-headers: "Authorization"
    nginx.ingress.kubernetes.io/auth-response-headers-prefix: "upstream_http_authorization"
spec:
  ingressClassName: nginx
  rules:
    - host: dittofeed.com
      http:
        paths:
          - path: /dashboard
            pathType: Prefix
            backend:
              service:
                name: dittofeed-cloud-dashboard
                port:
                  number: 80
          - path: /api
            pathType: Prefix
            backend:
              service:
                name: dittofeed-cloud-api
                port:
                  number: 80
```

## Service Env Vars

Environment variables to pass to Deployment resources of application.

### Dashboard Env Vars

put in deployment yaml file

```yaml
...
          env:
            - name: "AUTH_MODE"
              value: "multi-tenant"
            - name: "AUTH_PROVIDER"
              value: "auth0"
            - name: "OAUTH_START_URL"
              value: "https://dittofeed.com/oauth2/start?rd=https%3A%2F%2Fdittofeed.com%2Fdashboard"
            - name: "SIGNOUT_URL"
              value: "https://dittofeed.com/oauth2/sign_out?rd=https%3A%2F%2Fdittofeed.us.auth0.com%2Fv2%2Flogout%3FreturnTo%3Dhttps%253A%252F%252Fdittofeed.com%26client_id%3DO6FNPZkO6TJAWnAdb4cxpgoyWOuuzyM4"
```

### Worker and API Env Vars

```yaml
...
          env:
            - name: "AUTH_MODE"
              value: "multi-tenant"
            - name: "AUTH_PROVIDER"
              value: "auth0"
            - name: "OAUTH_START_URL"
              value: "https://dittofeed.com/oauth2/start?rd=https%3A%2F%2Fdittofeed.com%2Fdashboard"
```

**Comment by pcr0wley** [comment]
Potentially we could use Odoo as the auth provider however I still think that may require extra work.
https://odoo-community.org/shop/oauth-provider-9001
